### ISSUE TITLE
Breakpoints on SDT probes: "Attempt to write memory outside diversion session"

### ISSUE THREAD

**Issue by Bob131** [mozilla/rr#2406]
SystemTap SDT probes may be associated with a 'semaphore', a counter used to track whether a probe is watched. The probe can be guarded by a branch and is only triggered when a debugger has indicated its interest in it by incrementing the counter. This raises problems in an RR replay session as writes to memory are ignored.

An example session:
```
$ rr record ls
$ rr replay
(rr) b -probe libpthread:cond_init
(rr) c
Continuing.
[ERROR src/GdbServer.cc:514:dispatch_debugger_request()] Attempt to write memory outside diversion session
warning: Could not write the value of a SystemTap semaphore.
```

**Comment by rocallahan** [comment]
Are these implemented using uprobes? If so, did https://bugzilla.kernel.org/show_bug.cgi?id=203883 get fixed?

If they're not implemented using uprobes, can you point me at something describing how they're implemented?

**Comment by Bob131** [comment]
No, I believe they're unrelated. I've found documentation to be frustratingly scant, but there is some information to be found:
 - [SystemTap Wiki/UserSpaceProbeImplementation](https://sourceware.org/systemtap/wiki/UserSpaceProbeImplementation)
 - [A patch from LKML](https://lkml.org/lkml/2018/4/17/23) that contains a description of stap semaphores
 - [dtrace man page](https://sourceware.org/git/?p=systemtap.git;a=blob;f=man/dtrace.1;h=225f9301246ebeaa38e56288e8d7d25f22d629f6;hb=HEAD), also containing a section on stap semaphores
 - [sys/sdt.h](https://sourceware.org/git/?p=systemtap.git;a=blob;f=includes/sys/sdt.h;h=d1d10a658436de662600f7bc1bb061e731737ac6;hb=HEAD) header

**Comment by Bob131** [comment]
Oh, of course. I was still naively considering the case in which unguarded probe points are still associated with semaphores (like many in GLib).

> If there was a gdb remote protocol feature to support SDT in the agent directly, then probably something could be done in rr (which quite a bit of work, maybe using a read watchpoint on the semaphore value) to support them, but I don't think there is such a protocol feature.

I'm not sure I understand exactly what you mean. How I interpret that is
 - somehow be notified by gdb that a break was placed on a probe point
 - place a read watchpoint on the probe's semaphore
 - when the read watchpoint is triggered:
   - checkpoint the program state
   - create a diversion session to evaluate the alternate branch up to the probe point
   - trigger the probe breakpoint
   - restore the checkpoint

Is this understanding correct? If so, could rr detect semaphore adjustments by testing whether memory write addresses fall within `.probes` sections (which seems to be where the semaphores live, looking at the [dtrace source](https://sourceware.org/git/?p=systemtap.git;a=blob;f=dtrace.in;h=aa5f544a8954c3183e281fdcb9e15702669eef16;hb=HEAD))?

**Comment by rocallahan** [comment]
That is exactly what I was thinking, yes.

Mmm, OK, yeah, I guess that might work. I guess the diversion session would last until gdb hits a breakpoint, then we consider it done. But we'd have to stay in the diversion session when we report the breakpoint being hit, and exit the diversion automatically in the next continue, or something like that. Tricky but probably doable.

**Comment by rocallahan** [comment]
This is not something I plan to work on.

**Comment by Bob131** [comment]
I had started exposing functionality in `rr::Task` to allow querying ELF
sections by address, but part-way through I realised that rather than assuming
the use of dtrace it might be more robust to parse the ELF notes into a
structure. That is, more robust in the sense that probe semaphores not
contained in a `.probes` section would still be detected correctly. However,
it's possibly less robust in terms of interesting (ab)uses of notes (I'm
thinking things like semaphore addresses pointing into `.text` or unmapped
memory, or the program changing semaphore values under GDB's feet). Moreover,
I can't think of a scheme for validating stap notes (maybe it might make sense
for the purposes of some hack to overwrite code with semaphore values?) so it
seems like a potential minefield.

(I suppose that also raises questions about what the general strategy for
handling modification of global state by code guarded by a semaphore check
should be: the approach outlined previously would discard such modifications,
which strikes me as being Good Enough, but other people may have different
ideas)

Just wondering what your thoughts are regarding the better approach. FWIW, I'm
inclined to think that parsing ELF notes better comports with the stap design
documentation and how it's expected debugging tools will use them, accepting
that crazy hacks that violate assumptions aren't going to work everywhere.
That said, testing for section membership is probably simpler and the only
place I've seen semaphores not generated by dtrace is in the GDB test suite,
so that approach is probably perfectly adequate in practice.

As an aside, it turns out libpthread:cond_init doesn't actually have a
semaphore! Setting a breakpoint on it is trigging a bug in GDB in which the
test for absent semaphores isn't quite right. I've filed a patch to get that
fixed.

**Comment by Bob131** [comment]
Okay, here's a thought: for cases where code guarded by semaphore checks _do_
alter some non-transient program state (and a user needs this information),
maybe there could be some options for rr-record that accepted probe point
names for which semaphores would be incremented before recording? In such
cases, rr wouldn't have to handle any replay divergence at all.

In fact, maybe a better approach to the whole problem is just setting all
probe semaphores up-front in rr-record... I imagine rr would still have to
catch writes to semaphores during replay to avoid emitting spurious warnings,
but those writes could just be dropped.

**Comment by rocallahan** [comment]
Why don't you just record with an LD_PRELOAD library that sets semaphore values to something nonzero for the probes you care about? Then you don't have to modify rr at all.

**Comment by Bob131** [comment]
> Why don't you just record with an LD_PRELOAD library that sets semaphore
> values to something nonzero for the probes you care about?  Then you don't
> have to modify rr at all.

Because that doesn't resolve the main issue at hand, namely rr and GDB filling
the screen with warnings about divergent replay and errors about failed
semaphore writes respectively.

To be clear: I haven't yet seen any code that even uses semaphore values, so
ensuring that those values are set correctly is not actually part of my use
case. The walls of text above were less seeking a solution to my specific
problem (which is adequately solved by modifying rr to ignore divergent memory
writes and report success) and more an enquiry as to how a general solution
should look if it were to be accepted into upstream rr.

**Comment by rocallahan** [comment]
OK

**Comment by Bob131** [comment]
I have an extremely rough [proof-of-concept patch](https://github.com/mozilla/rr/files/3958119/patch.txt) that I've attached. Please have a look over it and let me know if this approach is appropriate.

The patch adds a new helper library, `librraudit`, used as an `LD_AUDIT` library to hook glibc's dynamic linker. For each loaded object it scans through the tags looking for SystemTap semaphores; when one is found, it's incremented and the address of the semaphore is signalled to rr via a new dummy syscall. If GDB attempts to write to one of these addresses during replay, the request is silently ignored.

**Comment by rocallahan** [comment]
A couple of nits:
* Doesn't build 32-bit audit library
* Doesn't handle case where `LD_AUDIT` is already set

My main concern is that this adds a bunch of complexity and some performance overhead for a case that's ... somewhat obscure.

**Comment by rocallahan** [comment]
How hard would it be to notify all the semaphores for an object in a single fake syscall?

What if we build the library but require users to manually load it by setting `LD_AUDIT` when recording?

What about library unloading?

**Comment by Bob131** [comment]
On Fri, Dec 13, 2019 at 07:55:05PM -0800, rocallahan wrote:
> How hard would it be to notify all the semaphores for an object in a single
> fake syscall?

...well, at the very least it should be pretty straight-forward to batch up
several semaphores in each syscall. AFAIK the semaphores are usually located
at consecutive addresses, so computing an interval covering semaphore
addresses and only invoking the syscall when a disjoint address is found
should lower the bandwidth a bit.

I'll have a closer look at the syscall buffering mechanism too, see if maybe
that could be used to send an array of addresses.

> What if we build the library but require users to manually load it by
> setting `LD_AUDIT` when recording?

I have no qualms with that, although maybe a command-line option would be
friendlier ;)

> What about library unloading?

Good point, I completely forgot about that. My initial reaction is to do a
similar thing: add an implementation of la_objclose() that iterates over the
object's semaphores again, invoking something like
SYS_rrcall_notify_stap_semaphore_byebye as it goes.

Thanks.

**Comment by rocallahan** [comment]
> I'll have a closer look at the syscall buffering mechanism too, see if maybe
that could be used to send an array of addresses.

It can't be used here, and it would just be simpler to batch together notifications manually.

I think we can land this if it's opt-in. If it were on by default I think that's a harder decision.

**Comment by Bob131** [comment]
On Fri, Dec 13, 2019 at 08:46:10PM -0800, rocallahan wrote:
> It can't be used here, and it would just be simpler to batch together
> notifications manually.

Ah, thanks for the heads up.

> I think we can land this if it's opt-in. If it were on by default I think
> that's a harder decision.

Sure. I'll submit a pull request once I've got everything in decent shape.

**Comment by Bob131** [comment]
Resolved by #2426
